cat/spirv.cat

```diff
@@ -0,0 +1,125 @@
+let viswg = [VIS & WG]; (chains & ((hb & ssg & avvisinc); vissg)?)
+let visqf = [VIS & QF]; (chains & ((hb & ssg & avvisinc) ; viswg)?; ((hb & ssg & avvisinc); vissg)?)
+let vissh = [VIS & SHADER]; (chains & ((hb & ssg & avvisinc) ; visqf)?; ((hb & swg & avvisinc); viswg)?; ((hb & sqf & avvisinc); vissg)?)
+let visdv = [VIS & DV]
+
+// loc order, relates memory accesses to the same location
+let locord = loc &
+             (  (hb & int & vloc) |
+                ([R \ PRIV]; hb; [R | W \ PRIV]) | // RaR, WaR (non-private)
+                ([R]; ssw+; [R | W]) | // RaR, WaR (any)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg);                          [W \ PRIV])) |   // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg); vissg; (po? & avvisinc); [R \ PRIV])) |   // RaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg);                          [W \ PRIV])) |   // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg); viswg; (po? & avvisinc); [R \ PRIV])) |   // RaW (via workgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf);                          [W \ PRIV])) |   // WaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf); visqf; (po? & avvisinc); [R \ PRIV])) |   // RaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);                                [W \ PRIV])) |   // WaW (via shader domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);       vissh; (po? & avvisinc); [R \ PRIV])) |   // RaW (via shader domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);                                [W]))        |   // WaW (via device domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);       visdv; (hb & avvisinc);  [R]))            // RaW (via device domain)
+             )
+
+let w-locord = [W]; locord
+
+let visto = (w-locord \ (w-locord; w-locord+)); [R] // visible to = location ordered W->R with no intervening write (W->W->R)
+(*******************)
+
+
+(****************)
+(* restrictions *)
+(****************)
+// rai[asmo] = ((A&W) -> (A&W)) & mutordatom
+empty ([ATOM & W]; mutordatom; [ATOM & W]) \ (asmo | asmo^-1) as coherence
+
+(***********************)
+(* Memory Model Axioms *)
+(***********************)
+// consistency: locord, rf, fr, asmo must not form cycles
+acyclic (locord | rf | fr | asmo) as consistency-cycle
+// consistency: non-atomic cannot read-from a value that is shadowed by another write
+empty (rf; ([R \ ATOM])) & (([W]; locord); ([W]; locord)+) as consistency-rf
+
+let read-write = (R \ W) * (R \ W) // exclude RMWs
+empty (((locord | locord^-1) | mutordatom \ read-write) \ (loc \ id \ read-write)) as locord-complete
```

Comment:
Can you check which is the order of precedence between `|` and `\` in alloy? I think suspect correct way of doing this would be `empty ((((locord | locord^-1) | mutordatom) \ read-write) \ (loc \ id \ read-write)) as locord-complete` and this changes fixes 4 more unit tests

Reply:
Fixed.
`|` and `\` hold the same precedence order in [alloy](http://alloytools.org/spec.html), but `\` is higher than `|` in cat. Hence the brackets is necessory.